forage-android/src/main/java/com/joinforage/forage/android/core/services/telemetry/Metrics.kt

```diff
@@ -72,71 +73,41 @@ internal enum class EventName(val value: String) {
     }
 }
 
-internal interface PerformanceMeasurer {
-    fun start()
-    fun end()
-    fun logResult()
-}
-
-internal interface NetworkMonitor : PerformanceMeasurer {
-    fun setPath(path: String): NetworkMonitor
-    fun setMethod(method: String): NetworkMonitor
-    fun setHttpStatusCode(code: Int): NetworkMonitor
-    fun setForageErrorCode(errorCode: String): NetworkMonitor
-}
-
-internal abstract class ResponseMonitor(metricsLogger: Log? = Log.getInstance()) : NetworkMonitor {
-    private var startTime: Long? = null
-    private var endTime: Long? = null
+internal abstract class ResponseMonitor<T>(metricsLogger: Log? = Log.getInstance()) {
+    private var startTime: Long
 
     private var logger: Log? = null
     private var responseAttributes: MutableMap<String, Any> = mutableMapOf()
 
     init {
         logger = metricsLogger
         responseAttributes[MetricsConstants.LOG_TYPE] = LogType.METRIC
-    }
-
-    override fun start() {
```

Comment:
If we made `metricsLogger` a parameter in a function call, it would make it trickier to accurately trace the `end - start` if we removed the `start()` method

---


```kotlin
fun someOperation(metricsLogger = newMeasurement()) {
  someSlowOperationInADifferentChain(...)

  tracedOperation()
  metricsLogger.finishAndReportResult()

  // now the response_time is affected by a method that we're not interested in tracing :(
}


vs. with `start`
fun someOperation(metricsLogger) {
  someOtherSlowOperation(...)

  metricsLogger.start()
  tracedOperation()
  metricsLogger.finishAndReportResult()
}
```

Are you comfortable with this tradeoff?

Reply:
This is a really thoughtful call. I have a few thoughts on this:
1. So far, we've not encountered a situation where we wanted to configure/instantiate a metric and delay starting the clock until another point in the code. Until it becomes more clear that behavior `.start()` or `.end()` is necessary or even nice to have, I think we get more value out of keeping our code as simple as possible
2. This decision is pretty reversible so if our code requirements evolve, we can absolutely bring back `.start()` and `.end()`

Thank you again for posing this tradeoff so thoughtfully

Reply:
> we've not encountered a situation where we wanted to configure/instantiate a metric and delay

We recently encountered this situation [in the JS SDK](https://github.com/teamforage/fe-monorepo/blob/e4d124efc5502ee49a805b94d585daec3362da87/apps/custom-checkout/src/utils/ForageSDKWrapper.ts#L208)

Making `measurement` a param can simplify testing (support DI)  and make it clear to the caller that a function is being "traced"/depends on the performance monitor.

In any case, I agree that it's easily-reversible, so we can revisit this  thread when the use case re-surfaces.